src/library/scala/collection/Seq.scala

```diff
@@ -891,9 +891,16 @@ trait SeqOps[+A, +CC[_], +C] extends Any
     *  @param  elem   the replacing element
     *  @tparam B        the element type of the returned $coll.
     *  @return a new $coll which is a copy of this $coll with the element at position `index` replaced by `elem`.
-    *  @throws IndexOutOfBoundsException if `index` does not satisfy `0 <= index < length`.
+    *  @throws IndexOutOfBoundsException if `index` does not satisfy `0 <= index < length`. In case of a
+    *                                    lazy collection this exception may be thrown at a later time or not at
+    *                                    all (if the end of the collection is never evaluated).
     */
-  def updated[B >: A](index: Int, elem: B): CC[B] = iterableFactory.from(new View.Updated(this, index, elem))
+  def updated[B >: A](index: Int, elem: B): CC[B] = {
+    if(index < 0) throw new IndexOutOfBoundsException(index.toString)
```

Comment:
Perhaps we could unify the two exception cases, and avoid having a structure-breaking (goto-like) `throw` using the following:  
```Scala
if (index < 0 ||  0 <= knownSize && knownSize <= index) 
   throw new IndexOutOfBoundsException(index.toString)
else 
   iterableFactory.from(new View.Updated(this, index, elem))
```

Reply:
`k >=0` is needed - it means that the size is known.

Reply:
@diesalbla while we could unify them, it would be better not to evaluate `knownSize` twice (while it is sub-linear, there's no guarantee that it's effectively a free call). you could still potentially inline a block,

```scala
if (index < 0 ||  { val k = knownSize; 0 <= k && k <= index }) 
   throw new IndexOutOfBoundsException(index.toString)
else 
   iterableFactory.from(new View.Updated(this, index, elem))
```

I'm not sure that improves readability at that point.